Dirichlet characters

b-mehta opened the pull request:
- [x] Create `M.NT.Moduli.Internal` module, for `discreteLogarithmPP` and `isPrimitiveRoot'`, so we can work with these internally without needing to use `CyclicGroup`.
- [x] I created `Template` a while ago, but since then `SFactors` was introduced - this can be adapted to make a nicer version of `Template`. It's unclear if `Template` should have the modulus on type level though, considering `induced` might be easier to write without that. Edit: This turned out to be a bad idea, due to the reason mentioned.
- [x] Figure out how to do `isPrimitive`
- [x] Bulk evaluation test currently fails. The [branch here](https://github.com/b-mehta/arithmoi/tree/dirichlet-characters-wip) might have the fix for this.

b-mehta committed:
Explicit import lists

b-mehta committed:
Generalise CRT and make canonical generators

b-mehta committed:
Fix crt underflow problems

b-mehta committed:
Initial testing

b-mehta committed:
Merge remote-tracking branch 'upstream/master' into dirichlet-characters

b-mehta committed:
Roots of unity

b-mehta committed:
Dirichlet characters first draft

b-mehta committed:
Minor corrections

b-mehta committed:
Merge remote-tracking branch 'upstream/master' into dirichlet-characters

b-mehta committed:
Fix merging problems

b-mehta committed:
(WIP) More dirichlet characters functionality

b-mehta committed:
Merge remote-tracking branch 'upstream/master' into dirichlet-characters

b-mehta committed:
Cosmetic changes

b-mehta committed:
Added Bounded instance and generalise conversions

b-mehta committed:
Renamed conversion functions

b-mehta committed:
Better dirichlet character tests

b-mehta committed:
More sensible Eq and renamed trivial

b-mehta committed:
Handling of real dirichlet characters

b-mehta committed:
Added more dirichlet tests

b-mehta committed:
Orthogonality relations and real tests

b-mehta committed:
No need for TypeLits.Mod

b-mehta committed:
Induce dirichlet characters to higher moduli

b-mehta committed:
Improved haddock for dirichlet characters

b-mehta committed:
Some better haddocks

b-mehta committed:
More sensible Eq and Enum instances for DChars

b-mehta committed:
Bitshifts instead of powers and cosmetic changes

b-mehta committed:
Merge remote-tracking branch 'upstream/master' into dirichlet-characters

b-mehta committed:
More internal documentation and haddocks

b-mehta committed:
valid character checker

b-mehta committed:
Avoid lots of recalculations for Enum instances

b-mehta committed:
Better jacobi (+tests), avoided some recalcuations

b-mehta committed:
simpler representation of characters

b-mehta committed:
Shuffle internals so we don't export discreteLogPP

b-mehta committed:
Added allChars, and switched twoPowers to Int

b-mehta committed:
Dirichlet characters are a group

b-mehta committed:
Allow evaluating a dirichlet character everywhere at once

b-mehta committed:
OrZero elsewhere and complete allEval

b-mehta committed:
Remove bad primitive character and more haddocks

b-mehta committed:
Merge branch 'master' into dirichlet-characters

b-mehta committed:
update to match new Mods

b-mehta committed:
add a new test

b-mehta committed:
cosmetic changes and comments

b-mehta committed:
Start of tests fix [skip ci]

Bodigrim commented:
This looks awesome! Extra kudos for beautiful TeX formulas.

b-mehta committed:
Correct implementation of induced characters

b-mehta committed:
Fix primitive testing

b-mehta committed:
Attempted bug fix

b-mehta committed:
attempted bug fix 2

Bodigrim commented:
(Just random notes)
At some point we would probably need a separate module dedicated to the roots of unity. There are `Moebius` and `JacobiSymbol`, which effectively are just `OrZero (RootsOfUnity 2)`, cubic residue symbols (#155) are `OrZero (RootsOfUnity 3)`, and units of Gaussian and Eisenstein integers are `RootsOfUnity 4` and `RootsOfUnity 6`.

Bodigrim commented:
BTW `OrZero a` seems to be equivalent to `Data.Monoid.Ap Maybe a`. I would rather stick with the latter, possibly adding a synonym `type OrZero = Ap Maybe`.

b-mehta commented:
> At some point we would probably need a separate module dedicated to the roots of unity. 

Yeah this is a good point, but I'm hesitant to make the RootOfUnity here be parametrised on type level. In particular, it would make
```haskell
evaluate :: DirichletCharacter n -> MultMod n -> RootOfUnity (Phi n)
```
which would be a pain (I did try). Perhaps `Some RootsOfUnity` would be better, but it seems a little awkward to work with.

Bodigrim commented:
Sure, let us leave `RootOfUnity` as is, at least for now.

b-mehta commented:
> BTW OrZero a seems to be equivalent to Data.Monoid.Ap Maybe a. I would rather stick with the latter, possibly adding a synonym type OrZero = Ap Maybe.

Ah, thanks for this! I'll convert to this, possibly using some pattern synonyms as well

b-mehta commented:
Actually, `Data.Monoid.Ap` was introduced in base 4.12, which would be a fairly strong new constraint... Should I still use it, or is it best to use `OrZero` for now, and create an issue to convert to `Data.Monoid.Ap` in the future?

b-mehta committed:
Misc fixups

Bodigrim commented:
I'd use `Data.Monoid.Ap` and define it for earlier versions of `base` along these lines:

```haskell
#if !MIN_VERSION_base(4,12,0)
newtype Ap f a = Ap { getApp :: f a }
  deriving (Functor, Applicative, Monad)

instance (Applicative f, Semigroup a) => Semigroup (Ap f a) where
  (<>) = liftA2 (<>)

instance (Applicative f, Monoid a) => Monoid (Ap f a) where
  mempty = pure mempty
  mappend = (<>)
#endif 
```

b-mehta committed:
More primitive chars API and real eval test

b-mehta committed:
remove a type application

b-mehta committed:
Use Monoid.Ap instead of OrZero

b-mehta committed:
Create internal module

b-mehta committed:
try again to get it to compile

b-mehta committed:
try again again to get it to compile

b-mehta committed:
try again again again to get it to compile

b-mehta committed:
[skip ci] Clean up TODOs

b-mehta commented:
I need to write a quick test for `orderChar`, but other than that I think this is essentially ready to review now. There's a few TODOs in there, but they can be separate issues so that this is easier to review.

b-mehta commented:
Do you have a preference for a Show instance for `DirichletCharacter`? The options as I see them are:

- None
- "χ_r with modulus n"
- Enumerate its values
- The internal representation
- Another option?

It's also tempting to implement `OrZero` with newtypes, so that a custom Show instance can be given, do you think this is a good idea?

b-mehta committed:
order test and remove comment

b-mehta committed:
Hlint suggestions

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+-- Licence:     MIT
+-- Maintainer:  Bhavik Mehta <bhavikmehta8@gmail.com>
+-- Stability:   Provisional
+-- Portability: Non-portable (GHC extensions)
```
I usually ignore `Stability` and `Portability`, IMHO they do not make much sense nowadays.

b-mehta replied:
Fair enough, I will remove.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE CPP #-}
```
Please sort pragmas alphabetically and justify right ends.

Bodigrim commented on arithmoi.cabal:
```diff
@@ -59,13 +59,15 @@ library
     Math.NumberTheory.Moduli.Class
     Math.NumberTheory.Moduli.DiscreteLogarithm
     Math.NumberTheory.Moduli.Equations
+    Math.NumberTheory.Moduli.Internal
```
I believe internals are not intended to be publicly exposed. Please move them under `other-modules` section.

b-mehta replied:
Thanks for this catch!

Bodigrim commented on Math/NumberTheory/Moduli/JacobiSymbol.hs:
```diff
@@ -48,6 +49,18 @@ negJS = \case
+                                   JacobiSymbol -> Natural
+  #-}
+-- | Convenience function to convert out of a Jacobi symbol
+symbolToIntegral :: Integral a => JacobiSymbol -> a
```
Probably `Num` constraint would be enough.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+-- >>> chi2 = induced @135 chi
+-- >>> :t chi2
+-- Maybe (DirichletCharacter 135)
+induced :: forall n d. (KnownNat d, KnownNat n) => DirichletCharacter d -> Maybe (DirichletCharacter n)
```
If you add `Mod n d ~ 0` to constraints, it would be possible to avoid `Maybe` at all.

b-mehta replied:
It would, but it would be harder to work with - even one of the tests would need to prove `Mod (d * n) d ~ 0` (which isn't in `Data.Constraint`). I think it's best to use `Divides d n` from Data.Constraint, and add the lemma `Divides a c => Divides a (b * c)`, but this would belong in a different branch as it's a bit of work.

Bodigrim replied:
I see. How do you feel about this?
```haskell
induced :: (KnownNat d, KnownNat n) => DirichletCharacter d -> DirichletCharacter (d * n)
```

b-mehta replied:
I also tried this; it requires `AllowAmbiguousTypes` in the module, and `TypeApplications` for the user: if `chi :: DirichletCharacter 5` then GHC can't figure out what `d` should be to make `induced chi :: DirichletCharacter 15` work, since `*` is non-injective. The user would instead need to do `induced @3 chi :: DirichletCharacter 15`, at which point we might as well use the version here, as far as I can see.

Bodigrim replied:
Yeah, I forgot that GHC treats `*` as non-injective because of `0`.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+        primitive (TwoPower 2 a _) = a /= mempty
+        primitive (TwoPower k _ b) = (bit (k-3) :: Integer) `stimes` b /= mempty
+
+-- | A Dirichlet character is primitive if cannot be @induced@ from any character with
```
Should be `'induced'`, not `@induced@`.

b-mehta replied:
Amended, thanks.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+  , evaluate
+  , generalEval
+  , toFunction
+  , allEval
```
Let us use the same stem: either `eval`, `generalEval`, `allEval` or `evaluate`, `generalEvaluate`, `allEvaluate`. I would also put the verb first: e. g., `evalAll` and `evalGeneral`.

b-mehta replied:
Adjusted, thanks.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+        two = toEnum 1 -- lazy way to get Prime 2
+
+-- | Get the order of the Dirichlet Character.
+-- TODO: test this
```
Remove `TODO`.

b-mehta replied:
Thanks.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+        validValued (OddPrime (unPrime -> p) k _ a) = (p^(k-1)*(p-1)) `stimes` a == mempty
+        validValued Two = True
+        n = natVal (Proxy :: Proxy n)
+        two = toEnum 1 -- lazy way to get Prime 2
```
Better use `minBound` :) 
It is actually an oversight from my side that `toEnum 1` is 2 and not 3.

b-mehta replied:
`minBound` doesn't work here because we would like a `Prime Natural`, which is not `Bounded`.

Bodigrim replied:
Ahh, I see. What about `nextPrime 2`?

b-mehta replied:
I switched to this, since I had no strong preference either way.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+  , allEval
+  -- ** Special Dirichlet characters
+  , principalChar
+  , isPrincipal
```
Isn't it redundant to export `isPrincipal`, when we have `Eq` instance already available? I would rather keep the export list as laconic as practical.

b-mehta replied:
Not really. I think your idea is that `isPrincipal chi` should be equivalent to `chi == mempty`, but `mempty` requires a `KnownNat n` instance, while `isPrincipal` does not. 
(`mempty` should require this instance, since to create a character from thin air, we need to know the modulus on term level)

Bodigrim replied:
Makes sense, thanks.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+  , toComplex
+  -- * An absorbing semigroup
+  , OrZero, pattern Zero, pattern NonZero
+  , asNumber
```
I would probably do not export this function. It is just `maybe 0`, not worth of polluting global namespace with a rather ambiguous name.

b-mehta replied:
It's `maybe 0` but with the newtype abstraction `Ap` in the way, so this would be useful to export. I'll happily take better name suggestions.

Bodigrim replied:
Ah, I am terrible with names. Probably `orZeroToNum` will do, if you still prefer it to `coerce (maybe 0)`.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,589 @@
+          d = denominator (2*q)
+
+-- | Given a rational \(q\), produce the root of unity \(e^{2 \pi i q}\).
+toRootOfUnity :: Integral a => Ratio a -> RootOfUnity
```
I'd rather avoid excessively polymorphic functions. Each call of `toRootOfUnity (2 % 3)` raises `Defaulting the following constraints to type ‘Integer’` warning. 

If you cast an argument to `Rational` first thing and do not use type `a` in any way, then it is better to make `toRootOfUnity` monomorphic, accepting only `Rational`. Users can easily combine it with `toRational` themselves, when needed.

b-mehta replied:
Changed, thanks

Bodigrim commented:
> Do you have a preference for a Show instance for `DirichletCharacter`? 

I would probably just dump its internal representation, but no strong opinion here.

> It's also tempting to implement `OrZero` with newtypes, so that a custom Show instance can be given, do you think this is a good idea?

Not realy. Too many layers and new newtypes make users struggle. However, you can probably define `instance {-# OVERLAPPING #-} Show (Ap (Maybe RootOfUnity))`.

b-mehta commented:
I would consider using overlapping instances significantly worse, since this extension can lead to incoherence. I think I'll just avoid a Show instance altogether.

b-mehta committed:
Changes from review

b-mehta committed:
A bunch more tests

b-mehta committed:
restore applicative import

Bodigrim commented:
> Yeah this is a good point, but I'm hesitant to make the RootOfUnity here be parametrised on type level. In particular, it would make
>
> ```haskell
> evaluate :: DirichletCharacter n -> MultMod n -> RootOfUnity (Phi n)
> ```
>
> which would be a pain (I did try).

Actually I think it is possible to avoid `Phi` along these lines:

```haskell
-- | RootOfUnity m :: RootOfUnity n represents \(e^{2 \pi i m / \phi(n)}\).
newtype RootOfUnity n = RootOfUnity Integer
  deriving (Eq)

instance KnownNat n => Show (RootOfUnity n) where
  show (RootOfUnity m) = "e^(" ++ show (numerator r) ++ "πi/" ++ show (denominator r) ++ ")"
    where
      r = 2 * m % totient (natVal (Proxy :: Proxy n))

toComplex :: Floating a => SFactors Integer n -> RootOfUnity n -> Complex a
toComplex fact (RootOfUnity m) = cis (2 * pi * m / d)
  where
    d = runFunctionOnFactors totientA (unSFactors fact)

evaluate :: DirichletCharacter n -> MultMod n -> RootOfUnity n
evaluate = undefined
```

(I do not mean this to be done in this branch)

Bodigrim commented on Math/NumberTheory/Moduli/Internal.hs:
```diff
@@ -0,0 +1,162 @@
+-- |
+-- Module:      Math.NumberTheory.Moduli.Multiplicative
+-- Copyright:   (c) 2017 Andrew Lelechenko
```
Please update `Module` and `Copyright` fields.

Bodigrim replied:
I meant that IIRC the copyright should be yours, not mine ;)

b-mehta replied:
Ah sure! :)

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,588 @@
+toRootOfUnity :: Rational -> RootOfUnity
+toRootOfUnity q = RootOfUnity ((n `rem` d) % d)
+  where n = toInteger $ numerator q
+        d = toInteger $ denominator q
```
Both casts are redundant.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,588 @@
+-- | Convert a Dirichlet character to a complex-valued function. As in `toComplex`, the result is
+-- inexact due to floating-point inaccuracies. See `toComplex`.
+toFunction :: (Integral a, RealFloat b, KnownNat n) => DirichletCharacter n -> a -> Complex b
+toFunction chi = asNumber toComplex . evalGeneral chi . fromIntegral
```
Similar to discussion about `toRootOfUnity`, please make this function monomorphic in `a`, accepting `Mod n` only.

b-mehta replied:
Do you think it is worth keeping this function? I'd put it in originally for flexibility, but it is just a composition of other exposed functions, so perhaps it doesn't belong.

Bodigrim replied:
Let's remove it then.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,588 @@
+-- | Give a collection of dirichlet characters from their numbers. This may be more efficient than
+-- `indexToChar` for multiple characters, as it prevents some internal recalculations.
+indicesToChars :: forall n a f. (KnownNat n, Integral a, Functor f) => f a -> f (DirichletCharacter n)
+indicesToChars = fmap (Generated . unroll t . (`mod` m) . fromIntegral)
```
Again, just consume `f Natural` instead of `f a` and eliminate `fromIntegral`.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,588 @@
+-- | Use one character to make many more: better than indicesToChars since it avoids recalculating
+-- some primitive roots
+bulkMakeChars :: (Integral a, Functor f) => DirichletCharacter n -> f a -> f (DirichletCharacter n)
+bulkMakeChars x = fmap (Generated . unroll t . (`mod` m) . fromIntegral)
```
Again, just consume `f Natural` instead of `f a` and eliminate `fromIntegral`.

b-mehta replied:
This function is not exposed but is used internally in multiple contexts, so if it's okay I'll leave this one polymorphic.

Bodigrim replied:
OK

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,588 @@
+-- inside each of the factor groups?
+-- | Evaluate a real Dirichlet character, which can only take values \(-1,0,1\).
+toRealFunction :: (Integral a, KnownNat n) => RealCharacter n -> a -> Int
+toRealFunction (RealChar chi) m = case evalGeneral chi (fromIntegral m) of
```
Please limit `a` to `Mod n` and eliminate `fromIntegral`.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,588 @@
+import Data.Monoid                                         (Ap(..))
+#endif
+import Data.Proxy                                          (Proxy(..))
+import Data.Ratio                                          (Rational, (%), numerator, denominator)
```
I'm getting a warning that `The import of ‘Rational’ from module ‘Data.Ratio’ is redundant`.

Bodigrim commented on Math/NumberTheory/DirichletCharacters.hs:
```diff
@@ -0,0 +1,588 @@
+  , validChar
+  ) where
+
+import Control.Applicative                                 (liftA2)
```
Please guard it using `CPP`:

```haskell
#if !MIN_VERSION_base(4,12,0)
import Control.Applicative 
#endif 
```

Bodigrim commented:
Overall a very nice code. I really appreciate the amount of work and engineering.

b-mehta commented:
> > Yeah this is a good point, but I'm hesitant to make the RootOfUnity here be parametrised on type level. In particular, it would make
> > ```haskell
> > evaluate :: DirichletCharacter n -> MultMod n -> RootOfUnity (Phi n)
> > ```
> > 
> > 
> > which would be a pain (I did try).
> 
> Actually I think it is possible to avoid `Phi` along these lines:
> 
> ```haskell
> -- | RootOfUnity m :: RootOfUnity n represents \(e^{2 \pi i m / \phi(n)}\).
> newtype RootOfUnity n = RootOfUnity Integer
>   deriving (Eq)
> 
> instance KnownNat n => Show (RootOfUnity n) where
>   show (RootOfUnity m) = "e^(" ++ show (numerator r) ++ "πi/" ++ show (denominator r) ++ ")"
>     where
>       r = 2 * m % totient (natVal (Proxy :: Proxy n))
> 
> toComplex :: Floating a => SFactors Integer n -> RootOfUnity n -> Complex a
> toComplex fact (RootOfUnity m) = cis (2 * pi * m / d)
>   where
>     d = runFunctionOnFactors totientA (unSFactors fact)
> 
> evaluate :: DirichletCharacter n -> MultMod n -> RootOfUnity n
> evaluate = undefined
> ```
> 
> (I do not mean this to be done in this branch)

This would work but it would mean RootOfUnity isn't the ideal name and we cannot represent cubic residues... I attempted a few other versions of this module which would be more typed, but they turned out to be awkward to use - you can see some attempts in my wip branch. I opted instead for a version without putting much on type level, but with tests that it all works (using validChar) instead

b-mehta committed:
changes from review

b-mehta committed:
Remove unnecessary definitions

b-mehta committed:
Other changes from review

b-mehta committed:
Change name to orZeroToNum

b-mehta committed:
construction from table

b-mehta committed:
simplify fromTable

b-mehta committed:
fix compile error

b-mehta committed:
Use inbuilt functions and add docs

b-mehta commented:
I've added `fromTable`, which allows one to construct a `DirichletCharacter` from its table of values, if a valid character exists, and also I've adjusted `iterateMaybe` to use `unfoldr` as opposed to manual list creation.

Here's a list of improvements that can be turned into issues after merge.

- `fromTable` currently requires the whole table, but only uses part of it to get the underlying character (and we verify the generated character is correct). In fact, it suffices to only supply a very small number of values to uniquely specify a character. If one specifies `chi` at a collection of points which generates `(Z/nZ)*`, this would uniquely define the character (although this might be hard to do in practice). With this in mind, we can make `fromTable` more general:
```haskell
data DirichletError = UnderSpecified | OverSpecified | InvalidValues
fromTable :: forall n. KnownNat n => Vector (OrZero RootOfUnity) -> Either DirichletError (DirichletCharacter n)
```
- `toRealFunction` uses `evalGeneral`, but it should be possible to calculate it without doing any discrete log calculations - I believe we can use `jacobiSymbol` to do the odd prime power cases for instance - which would make it more efficient
- (Easy) Conversion to/from the right sort of ArithmeticFunction (the conversion `from` ought to be partial, using `fromTable`) - this was mentioned in the original issue. If it's still desired, the conversion `from` would need an unchecked precondition that the given function is periodic. The conversion `to` should use `evalAll` since we might need to evaluate it at every prime, which cover almost all of `MultMod n`.

The following aren't directly related to characters, but they came up while working on this.

- It would be nice to have a generalised version of `chineseCoprime` which works on more than two cases
- It might be useful to export
```haskell
reduceMod :: forall a b. (KnownNat a, KnownNat b) => Mod (a*b) -> (Mod a, Mod b)
reduceMod (Mod n) = (fromIntegral n, fromIntegral n)

reduceMult :: forall a b. (KnownNat a, KnownNat b) => MultMod (a*b) -> (MultMod a, MultMod b)
reduceMult (MultMod n) = (MultMod a, MultMod b)
  where (a,b) = reduceMod n
```

b-mehta committed:
remove redundant pragma

Bodigrim commented:
Great job, thanks for your efforts and time!

----

> `fromTable` currently requires the whole table, but only uses part of it 

I'd probably go for 

```haskell
fromTable :: forall n. KnownNat n => (Mod n -> OrZero RootOfUnity) -> DirichletCharacter n
fromTable f = undefined
```

such that `fromTable` evaluates `f` on each group generator and reconstructs all the rest. And `f` may lookup into a vector or whatever it likes.

> `toRealFunction` uses `evalGeneral`, but it should be possible to calculate it without doing any discrete log calculations

That's a good point.

> Conversion to/from the right sort of `ArithmeticFunction`

I think there are enough helpers for a user willing to implement it himself, so this is of a low priority probably.

> It would be nice to have a generalised version of chineseCoprime which works on more than two cases

What exactly do you have in mind? 

> It might be useful to export `reduceMod`

This could make a nice addition to https://github.com/Bodigrim/mod

b-mehta commented:
> such that fromTable evaluates f on each group generator and reconstructs all the rest. And f may lookup into a vector or whatever it likes.

I don't think this makes much of a difference; besides `fromTable` still checks that the character is valid, which is done by ensuring the whole table matches.
What I had in mind was more that some values can be unspecified while still giving the character - but it would need a more advanced method than the one I used here.

> What exactly do you have in mind?

This is what I'd quickly guessed earlier, though I don't doubt something better exists
```haskell
chineseCoprimeList :: (Eq a, Ring a, Euclidean a) => [(a, a)] -> Maybe a
chineseCoprimeList xs = fmap fst $ foldM (\x y -> fmap (,snd x `times` snd y) (chineseCoprime x y)) (zero, one) xs
```
